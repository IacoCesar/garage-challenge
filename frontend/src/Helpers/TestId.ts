import { Platform } from 'react-native'

interface ITestID {
  testID: string
  accessibilityLabel?: string
  accessible?: boolean
}

type componentType =
  | 'TouchableOpacity'
  | 'View'
  | 'SafeAreaView'
  | 'List'
  | 'Modal'
  | 'Text'
  | 'Image'
  | 'ActivityIndicator'

const testID = (testID: string, accessibilityLabel?: string): ITestID =>
  Platform.OS === 'android'
    ? {
        testID,
        accessibilityLabel,
        accessible: true
      }
    : {
        testID,
        accessible: false
      }

export const assignTestId = (
  componentType: componentType,
  componentId: string,
  componentText?: string
): ITestID => {
  const id = `${componentType}_${componentId}`

  return testID(id, componentText)
}

export const mockTestId = (
  componentType: componentType,
  componentId: string,
  componentText?: string
): ITestID => {
  return assignTestId(componentType, componentId, componentText)
}
