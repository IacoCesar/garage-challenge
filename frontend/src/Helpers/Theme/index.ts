import { DefaultTheme } from 'styled-components/native'

export const commonThemes = {
  colors: {
    bgWhite: '#EFEDEE',
    white: '#FFFF',
    black: '#000000',
    cardColor: '#ffffff',
    starColor: '#F9E26B',
    divisorColor: '#9F9F9F'
  },
  spacing: {
    global: {
      horizontal: 16
    }
  }
}

export const theme: DefaultTheme = {
  standard: {
    darkMode: {
      icon: commonThemes.colors.white,
      divisor: commonThemes.colors.bgWhite,
      shadow: commonThemes.colors.white,
      card: commonThemes.colors.divisorColor,
      cardTextColor: commonThemes.colors.black,
      background: commonThemes.colors.black,
      textColor: commonThemes.colors.white,
      ...commonThemes
    },
    default: {
      icon: commonThemes.colors.black,
      divisor: commonThemes.colors.divisorColor,
      shadow: commonThemes.colors.black,
      card: commonThemes.colors.cardColor,
      cardTextColor: commonThemes.colors.black,
      background: commonThemes.colors.white,
      textColor: commonThemes.colors.black,
      ...commonThemes
    }
  },
  ...commonThemes
}
