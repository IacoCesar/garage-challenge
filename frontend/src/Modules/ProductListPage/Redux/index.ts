import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IVitrineState, IPayloadProps } from './types'

export const initialState: IVitrineState = {
  vitrineItemSelector: {
    id: '',
    model: '',
    make: '',
    year: 0,
    image: {
      url: ''
    }
  }
}

export const VitrineSlice = createSlice({
  name: 'vitrine',
  initialState,
  reducers: {
    setVitrineItemSelector: (
      state: IVitrineState,
      action: PayloadAction<IPayloadProps>
    ) => {
      state.vitrineItemSelector = action.payload
    }
  }
})

export const { setVitrineItemSelector }: any = VitrineSlice.actions

export default VitrineSlice.reducer
