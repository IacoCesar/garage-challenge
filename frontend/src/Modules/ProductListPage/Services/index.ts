import { API_BASE } from '@env'
import axios from 'axios'

const Services = {
  getVitrine: () => {
    return axios.get(`${API_BASE}/api/cars`)
  },
  getFavoriteCars: () => {
    return axios.get(`${API_BASE}/api/cars/favorite`)
  },
  favoriteCar: (params: object) => {
    return axios.post(`${API_BASE}/api/cars/favorite`, params)
  },
  unFavoriteCar: (params: object) => {
    return axios.post(`${API_BASE}/api/cars/unfavorite`, params)
  }
}

export default Services
