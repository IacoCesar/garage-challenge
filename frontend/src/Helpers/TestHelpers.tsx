import React from 'react'
import { SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components/native'
import { theme } from '@/Helpers/Theme'
import configureStore from 'redux-mock-store' //ES6 modules

import { render } from '@testing-library/react-native'

const mockStore = configureStore([])

const renderProviders = ({ children }: any): React.ReactElement => {
  const store = mockStore({
    state: { themeData: { theme: 'default' } }
  })

  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <SafeAreaProvider>{children}</SafeAreaProvider>
      </ThemeProvider>
    </Provider>
  )
}

const customRender = (ui: React.ReactElement, options?: object) =>
  render(ui, {
    wrapper: renderProviders,
    ...options
  })

export * from '@testing-library/react-native'

export { customRender as render }
