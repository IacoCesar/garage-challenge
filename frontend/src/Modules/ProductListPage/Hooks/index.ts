import { useCallback, useEffect, useState } from 'react'
import Services from '../Services'

export const useProductListPage = () => {
  const [isLoading, setIsLoading] = useState(false)
  const [vitrineData, setVitrineData] = useState([])
  const [vitrineDataControlled, setVitrineDataControlled] = useState([])
  const [vitrineDataCount, setVitrineDataCount] = useState(2)
  const [favoriteCarsData, setFavoriteCarsData] = useState([])

  useEffect(() => {
    if (vitrineDataControlled.length === 0) return

    callLoadVitrineDataCount()
  }, [vitrineDataControlled])

  const callLoadVitrineDataCount = useCallback(() => {
    if (vitrineData.length === vitrineDataControlled.length) return

    const filteredVitrineData = vitrineDataControlled.filter(
      (_, index) => index < vitrineDataCount
    )

    setVitrineData(filteredVitrineData)
    setVitrineDataCount((prev) => prev + 2)
  }, [vitrineDataControlled, vitrineDataCount, vitrineData])

  const callLoadVitrineData = useCallback(async () => {
    try {
      if (vitrineDataControlled.length > 0) return callLoadVitrineDataCount()

      setIsLoading(true)

      const { data, status } = await Services.getVitrine()

      if (status !== 200) return

      setVitrineDataControlled(data.items)
    } catch (error) {
      /// error here
    } finally {
      setIsLoading(false)
    }
  }, [vitrineData, vitrineDataControlled])

  const callFavoriteCar = useCallback(async (id: string) => {
    try {
      setIsLoading(true)

      const params = { carId: id }

      const { status } = await Services.favoriteCar(params)

      if (status !== 201) return

      callLoadFavoriteCars()
    } catch (error) {
      /// error here
    } finally {
      setIsLoading(false)
    }
  }, [])

  const callUnFavoriteCar = useCallback(async (id: string) => {
    try {
      setIsLoading(true)

      const params = { carId: id }

      const { status } = await Services.unFavoriteCar(params)

      if (status !== 200) return

      callLoadFavoriteCars()
    } catch (error) {
      /// error here
    } finally {
      setIsLoading(false)
    }
  }, [])

  const callLoadFavoriteCars = useCallback(async () => {
    try {
      const { data, status } = await Services.getFavoriteCars()

      if (status !== 200) return

      setFavoriteCarsData(data.items)
    } catch (error) {
      /// error here
    }
  }, [])

  return {
    isLoading,
    vitrineData,
    vitrineDataCount,
    favoriteCarsData,
    callLoadVitrineData,
    callFavoriteCar,
    callUnFavoriteCar,
    callLoadFavoriteCars
  }
}
