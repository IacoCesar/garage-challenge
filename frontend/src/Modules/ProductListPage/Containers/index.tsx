import React, { useEffect } from 'react'
import { TNavigation } from '@/Helpers/types'
import { useNavigation } from '@react-navigation/native'
import Vitrine from '../Components/Vitrine'
import { useProductListPage } from '../Hooks'
import { useDispatch } from 'react-redux'
import { setVitrineItemSelector } from '../Redux'

const ProductListPage: React.FC = () => {
  const navigation = useNavigation<TNavigation>()
  const dispatch = useDispatch()

  const {
    isLoading,
    vitrineData,
    favoriteCarsData,
    callLoadVitrineData,
    callFavoriteCar,
    callUnFavoriteCar,
    callLoadFavoriteCars
  } = useProductListPage()

  useEffect(() => {
    callLoadVitrineData()
    callLoadFavoriteCars()
  }, [])

  const callHandleOpenPdp = (vitrineItemData: object) => {
    dispatch(setVitrineItemSelector(vitrineItemData))

    navigation.navigate('ProductDetailPage')
  }

  return (
    <Vitrine
      isLoading={isLoading}
      vitrineData={vitrineData}
      favoriteCarsData={favoriteCarsData}
      callHandleOpenPdp={callHandleOpenPdp}
      callHandleFavoriteCar={callFavoriteCar}
      callHandleUnFavoriteCar={callUnFavoriteCar}
      callLoadVitrineData={callLoadVitrineData}
    />
  )
}

export default ProductListPage
