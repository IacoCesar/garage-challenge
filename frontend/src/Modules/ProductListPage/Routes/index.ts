import ProductListPage from '../Containers'

export default { ProductListPage }
