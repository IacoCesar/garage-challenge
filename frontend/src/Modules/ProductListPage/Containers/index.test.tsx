import React from 'react'
import ProductListPage from '.'
import { render } from '@/Helpers/TestHelpers'

describe('ProductListPage', () => {
  it('should render snapshot', async () => {
    const wrapper = render(<ProductListPage />)

    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
