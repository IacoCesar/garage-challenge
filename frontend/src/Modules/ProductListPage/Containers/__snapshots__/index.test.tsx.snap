// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListPage should render snapshot 1`] = `
Array [
  <View
    accessible={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.3)",
          "display": "flex",
          "height": "100%",
          "justifyContent": "center",
          "position": "absolute",
          "width": "100%",
          "zIndex": 5,
        },
      ]
    }
    testID="View_Vitrine_LoaderContent"
  >
    <ActivityIndicator
      accessible={false}
      animating={true}
      color="#FFFF"
      hidesWhenStopped={true}
      size={25}
      style={
        Array [
          Object {},
        ]
      }
      testID="ActivityIndicator_Vitrine_LoaderContent_Content_LoadingIndicator"
    />
  </View>,
  <RCTScrollView
    ListHeaderComponent={
      <React.Fragment>
        <SpacerView
          max={16}
          min={16}
          testID="Vitrine_Content_Spacer"
        />
        <Styled(FlatList)
          ItemSeparatorComponent={[Function]}
          data={Array []}
          keyExtractor={[Function]}
          onEndReached={[Function]}
          onEndReachedThreshold={0.5}
          renderItem={[Function]}
          testID="Vitrine_Content_List"
        />
        <SpacerView
          max={16}
          min={16}
          testID="Vitrine_Content_Spacer_2"
        />
      </React.Fragment>
    }
    accessible={false}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFF",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
    testID="List_Vitrine_Content"
    themeSelector="default"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          accessible={false}
          max={16}
          min={16}
          style={
            Array [
              Object {
                "flexGrow": 1,
                "maxHeight": 16,
                "minHeight": 16,
              },
            ]
          }
          testID="View_Vitrine_Content_Spacer"
        />
        <View
          ItemSeparatorComponent={[Function]}
          data={Array []}
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          onContentSizeChange={[Function]}
          onEndReached={[Function]}
          onEndReachedThreshold={0.5}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          style={
            Array [
              Object {
                "paddingHorizontal": 16,
              },
            ]
          }
          testID="Vitrine_Content_List"
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        />
        <View
          accessible={false}
          max={16}
          min={16}
          style={
            Array [
              Object {
                "flexGrow": 1,
                "maxHeight": 16,
                "minHeight": 16,
              },
            ]
          }
          testID="View_Vitrine_Content_Spacer_2"
        />
      </View>
    </View>
  </RCTScrollView>,
]
`;
